File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Optimization/Nodes/TerminationViaMaxIterations.scala
package OSOL.Extremum.Core.Scala.Optimization.Nodes

import OSOL.Extremum.Core.Scala.Optimization._
import OSOL.Extremum.Core.Scala.Optimization.Exceptions._

class TerminationViaMaxIterations[Base, FuncType, V <: Optimizable[Base, FuncType]](val maxIteration: Int, val nextNodeId: Int, override val nodeId: Int)
  extends GeneralNode[Base, FuncType, V ](nodeId) {

  final private val parameterName = "currentIteration"

  final override def initialize(f: Map[String, FuncType] => FuncType, area: Area, s: State[Base, FuncType, V]): Unit = {
    try {
      val _ = s.getParameter[Int](parameterName)
      throw new ParameterAlreadyExistsException(parameterName)
    }
    catch {
      case _: NoSuchParameterException => s.setParameter(parameterName, 0)
      case _ => throw new Exception("Unknown Exception")
    }
  }

  final override def process(f: Map[String, FuncType] => FuncType, area: Area, s: State[Base, FuncType, V]): Unit = {
    s.setParameter(parameterName, s.getParameter[Int](parameterName) + 1)
  }

  final override def getCurrentCondition(f: Map[String, FuncType] => FuncType, area: Area, s: State[Base, FuncType, V]): Option[Int] = {
    if (s.getParameter[Int](parameterName) > maxIteration) Some(-1)
    else Some(nextNodeId)
  }

}

File: Cores/OSOL.Extremum.Core.DotNet/Optimization/Nodes/TerminationViaMaxIterations.cs
using System;
using System.Collections.Generic;

namespace OSOL.Extremum.Core.DotNet.Optimization.Nodes
{
    public class TerminationViaMaxIterations<TBase, TFuncType, TV>: GeneralNode<TBase, TFuncType, TV>
        where TV: class, IOptimizable<TBase, TFuncType>
    {
        private string ParameterName;
        private int MaxIteration;

        public TerminationViaMaxIterations(int nodeId, int maxIteration, string parameterName = "currentIteration")
        {
            this.NodeId = nodeId;
            this.MaxIteration = maxIteration;
            this.ParameterName = parameterName;
        }

        public override void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state)
        {
            int? value = null;
            try
            {
                value = state.GetParameter<int>(ParameterName);
            }
            catch (OptimizationExceptions.NoSuchParameterException e)
            {
                state.SetParameter<int>(name: ParameterName, value: 0);
            }
            finally
            {
                if(value.HasValue) throw new OptimizationExceptions.ParameterAlreadyExistsException(ParameterName);
            }
        }

        public override void Process(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state)
        {
            state.SetParameter(name: ParameterName, value: state.GetParameter<int>(ParameterName) + 1);
        }

        public override int? GetCurrentCondition(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state) => 
            state.GetParameter<int>(ParameterName) > MaxIteration ? 1 : 0;

    }
}